metaflow/plugins/secrets/secrets_decorator.py

```diff
@@ -210,8 +210,17 @@ def task_pre_step(
         ubf_context,
         inputs,
     ):
-        if ubf_context == UBF_CONTROL:
-            """control tasks (as used in "unbounded for each") don't need secrets"""
+        # We will skip the secret injection for "locally" launched UBF_CONTROL tasks because otherwise there
+        # will be a env var clash with UBF_WORKER tasks which will fail the flow in the @secrets decorator's `task_pre_step`.
+        if (
+            ubf_context == UBF_CONTROL
+            and os.environ.get("METAFLOW_RUNTIME_ENVIRONMENT", "local") == "local"
+        ):
+            # When we "locally" run @parallel tasks, the control task will create the worker tasks and the environment variables
+            # of the control task are inherited by the worker tasks. If we don't skip setting secrets in the control task then the
+            # worker tasks will already have the secrets set as environment variables, causing the @secrets' `task_pre_step` to fail.
+            # In remote settings, (e.g. AWS Batch/Kubernetes), the worker task and control task are independently created
+            # so there is no chances of an env var clash.
```

Comment:
can we change the behavior of `@secrets` such that it is a no-op for worker tasks locally, instead of it being a no-op for control tasks? that ensures parity for remote and local execution

Reply:
If I make this change then we need the subsequent changes to the testing of the secrets deco with parallel in this PR otherwise the tests would fail.

Reply:
Done.